### ISSUE TITLE
Unused attribute

### ISSUE THREAD

**Issue by ivan-pi** [j3-fortran/fortran_proposals#316]
When passing procedures as callbacks, sometimes a subset of the expected procedure parameters is not required and goes unused. Compiler tends to emit (false) warnings for unused parameters, despite this being deliberate in such cases.

```fortran
  subroutine myode(n,t,y,yp,ptr)
    integer, value :: n 
    real(wp), intent(in) :: t, y(n)
    real(wp), intent(out) :: yp(n)
    type(c_ptr), value :: ptr  ! ptr is unused

    real(wp), pointer :: params(:)
    
    real(wp), parameter :: k1 = 0.1_wp, k2 = 0.05_wp
    yp(1) = -k1*y(1)
    yp(2) =  k1*y(1) - k2*y(2)
    yp(3) =  k2*y(2)
  end subroutine
```

In C++17, it is possible to suppress warnings on unused entities using the [`[[maybe_unused]]`](https://en.cppreference.com/w/cpp/language/attributes/maybe_unused) attribute:
```cpp
[[maybe_unused]] void f(int n, double t, double *y, double *dydt, 
        [[maybe_unused]] void *params)
```

Another convention to prevent inadvertent use in C/C++ is to leave the dummy arguments unnamed:
```
void f(int n, double t, double *y, double *dydt, 
        void * /* params */)
```

In MATLAB, one can also [ignore function inputs](https://www.mathworks.com/help/matlab/matlab_prog/ignore-function-inputs.html) using the `~` symbol:
```
function dydt = f(t,y,~)
```

Is there interest to have something similar in Fortran? An attributes syntax could be used for other purposes too.

**Comment by certik** [comment]
Good idea. Probably `type(c_ptr), value, unused :: ptr  ! ptr is unused` seems the most natural.

**Comment by ivan-pi** [comment]
The reason it is called `maybe_unused` in C++ is that due to preprocessor logic or template meta-programming, you can end up with situations where a variable appears to be used, but it gets cut out before the function reaches compilation. With a potential preprocessor in 202Y is this something for consideration here too?

**Comment by certik** [comment]
Can you show an example of this? Isn't the use case rather that a given variable can both be used and unused, depending on a preprocessor option? But it's never "maybe used" (some kind of superposition of used and unused?), it seems it is always either used or unused.

**Comment by FortranFan** [comment]
@ivan-pi wrote Oct. 6, 2023 05:37 PM EDT:
> .. Is there interest to have something similar in Fortran?  ..

For whatever it's worth, my firm opinion is something like "maybe_unused" does **not** fit in the language standard.

I think there are two parts to the underlying issue:

1. First is the immediate annoyance with the "warning"s issued by overeager compilers when asked to put on their pedantic hats.  To address this, what will be called for is the work those compiler implementors.  In the case of `LFortran`, @certik is setting a great example by championing a compiler/toolset ecosystem that works best for practitioners causing the least annoyance.  @certik thus establishes the paradigm, "where there is will, there is a way".  And the Community of `LFortran` can come up with best ideas together with @certik and team to take care of this.  In the case of Intel Fortran users, the workflow will be for those "annoyed" with it to post threads at Intel Fortran forum and get Intel staff to work with Fortran team to figure this out.  In the interim, a significant userbase uses a build system outside of `fpm` such as Visual Studio IDE and there are ways to suppress specific warning locally and that can mitigate the annoyance.  In the case of `gfortran`, the users are the implementors, whether they like it or not.  And the annoyed users need to then become compiler developers and sort this one out in their compiler itself.
2. Next is the longer term aspect with Generics.  Currently the Fortran standard does not include any semantics toward resolution of generic interfaces, meaning a procedure with `interface` A cannot be disambiguated relative to `interface` B.  This places, what I see as needless, limitations on library developers.  And in the case of callbacks, it leads to scenarios as shown by @ivan-pi , where some received arguments can be unused in user procedures for the callback.  In some cases, as an (erstwhile) library developer, I have found the `OPTIONAL` attribute to come in handy.  But it is not always the case.  I do think the J3 Generics subgroup, either directly or via supplemental direction to other subgroup such as /Data, should do the needful to ensure the language standard offers better.  I will elaborate in a subsequent comment.

**Comment by certik** [comment]
Another idea is to do what Rust does, which prefixes variable names with "_" to avoid unused warnings.

Thanks @FortranFan for your comments. Yes indeed, compiler cooperation will be key here.

**Comment by FortranFan** [comment]
> 2\. I will elaborate in a subsequent comment.

Consider this example:
```fortran
module m
   abstract interface
      function Ifunc1(x) result(r)
         integer, intent(in) :: x
         integer :: r 
      end function
      function Ifunc2(x, y) result(r)
         integer, intent(in) :: x
         integer, intent(in) :: y
         integer :: r 
      end function
   end interface
   interface solve
      module procedure solve1
      module procedure solve2
      module procedure solve3
   end interface
contains
   subroutine solve1(x, proc)
      integer, intent(inout) :: x
      procedure(Ifunc1) :: proc
      x = proc( x )
   end subroutine 
   subroutine solve2(x, y, proc)
      integer, intent(inout) :: x
      integer, intent(in) :: y
      procedure(Ifunc2) :: proc
      x = proc( x, y )
   end subroutine 
   subroutine solve3(x, proc)  !<-- indistinguishable from solve1
      integer, intent(inout) :: x
      procedure(Ifunc2) :: proc
      integer :: y
      y = 0
      x = proc( x, y )
   end subroutine 
end module 
```
```
C:\temp>gfortran -c -ffree-form m.f
m.f:19:20:

   19 |    subroutine solve1(x, proc)
      |                    1
......
   30 |    subroutine solve3(x, proc)
      |                    2
Error: Ambiguous interfaces in generic interface 'solve' for 'solve1' at (1) and 'solve3' at (2)
```
 Improved semantics in the language can take into account specific elements of the interface e,g., `Ifunc1` vs `Ifunc2` in the above example to resolve the generic interface `solve`.

**Comment by klausler** [comment]
> In the case of gfortran, the users are the implementors, whether they like it or know it or not! And the annoyed users need to then become compiler developers and sort this one out in their compiler itself.

Wait, what?  Why are GNU Fortran users all "implementors", but users of other compilers not?

**Comment by FortranFan** [comment]
> Why are GNU Fortran users all "implementors", but users of other compilers not?

`gfortran` can really benefit from more FOSS volunteers to advance the compiler such as with pending standard features from Fortran 2003 onward and to help resolve the growing list of open incidents in their GCC Bugzilla systems.

Users of `gfortran` themselves are welcomed to be such FOSS volunteers, especially if they seek more timely attention to matters of interest to them i.e., they can become the implementors.

**Comment by septcolor** [comment]
If this feature becomes available, I will definitely use it to suppress warnings about intentionally unused variables (because I always attach a check option to catch all unused variables).
```
type(Foo_t), unused :: foo
class(Myclass_t), unused :: this
```
To mimic this feature, I am using a CPP macro + associate to cheat compilers, like
```
#define _touch_(x) associate(dum__ => x); endassociate
```
and use it like `_touch_(foo)`.
